Added support for static private fields accessors and methods.

dragomirtitian opened the pull request:
Draft PR for private static fields methods and accessors.

dragomirtitian committed:
Added support for static private fields accessors and methods.

dragomirtitian committed:
Added error message for private identifiers used with static decorators. There is no spec to go with this behavior as of yet.

dragomirtitian committed:
Fixed emit static bug that used private names outside of classes for initialization in esnext. Fixed issue where nested privates produce incorrect brand check.

dragomirtitian committed:
Added tests for private static fields methods and accessors.

dragomirtitian committed:
Merge remote-tracking branch 'remotes/bloomberg/es-private-methods-and-accessors' into es-private-static-fields-methods-and-accessors

dragomirtitian committed:
Merge branch 'es-private-methods-and-accessors' into es-private-static-fields-methods-and-accessors

dragomirtitian committed:
Accept new baseline.

dragomirtitian committed:
Improved duplicate identifier checks for static private class elements.

joeywatts commented on tests/baselines/reference/privateNameStaticAccessors.js:
```diff
@@ -0,0 +1,44 @@
+//// [privateNameStaticAccessors.ts]
+class A1 {
+    static get #prop() { return ""; }
```
What if you had `static obj = new A1();` before this?

dragomirtitian replied:
Thank you for that comment, it resulted in a fix here for fields and two bugs for babel [this](https://github.com/babel/babel/issues/12905) and [this](https://github.com/babel/babel/issues/12904) 😊

joeywatts commented on src/compiler/transformers/classFields.ts:
```diff
@@ -937,6 +1064,12 @@ namespace ts {
+                        case PrivateIdentifierPlacement.StaticField: {
+                            return createPrivateStaticFieldInitializer(
+                                privateIdentifierInfo.variableName,
+                                visitNode(property.initializer, visitor, isExpression)
```
What if the initializer uses something like `this.#myFoo`? I think `this` is bound as the class when the initializer is being executed.

dragomirtitian replied:
@joeywatts using `this` in static initializers is forbidden by TS at the moment. There is another ticket that is winding it's way to solve this. I wouldn't add any extra scope to this PR to attempt to fix this for private fields only.

joeywatts reviewed (commented):
Couple of initial thoughts. Will try to look over again

mkubilayk reviewed (commented):
I think @joeywatts also mentioned this in one of his comments but the following example doesn't work correctly right now:

```ts
class C {
    static #x = 42;
    static #y = this.#z();
    static #z() {
        return this.#x;
    }

    static m() {
        return C.#y;
    }
}

console.log(C.m()); // should log 42
```

We should fix the checker (produces `TS2334`) and the emit (should probably replace `this` in the field initializers with the class constructor?).

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1341,6 +1544,15 @@ namespace ts {
+    function createPrivateStaticFieldInitializer(variableName: Identifier, initializer: Expression | undefined) {
+        return factory.createAssignment(
+            variableName,
+            factory.createObjectLiteralExpression([
```
Do we need this to be an object?

dragomirtitian replied:
Since we use a function to set the value of the static property and JS can't pass by reference, this was the simplest solution I could think of: Pass in the wrapper object to the static fields set function in order to allow it to mutate the value property.

dragomirtitian replied:
As we talked offline, we will stick with the container object.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1103,8 +1237,67 @@ namespace ts {
-
-            if (isPropertyDeclaration(node)) {
+            const env = getPrivateIdentifierEnvironment();
+            if(hasStaticModifier(node)) {
```
nit: There are some inconsistencies in styling `if` statements: `if ()` vs. `if()`.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1103,8 +1237,67 @@ namespace ts {
+                            ...previousInfo,
+                            placement: PrivateIdentifierPlacement.StaticGetterAndSetter,
+                            getterName,
+                            setterName: previousInfo.setterName,
```
I think this property comes from the spread already.

dragomirtitian replied:
Yup, Thanks.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1103,8 +1237,67 @@ namespace ts {
+                        classConstructor: env.classConstructor
+                    };
+                }
+                else if(isGetAccessorDeclaration(node)) {
```
I see instance accessors use different utilities: `isGetAcessor` vs. `isGetAccessorDeclaration`. What is the difference?

dragomirtitian replied:
This is really strange .. there are two functions with the exact same definition .. just from different lcoations:

```ts
    // src\compiler\utilitiesPublic.ts
    export function isGetAccessor(node: Node): node is GetAccessorDeclaration {
        return node.kind === SyntaxKind.GetAccessor;
    }

    //src\compiler\factory\nodeTests.ts
    export function isGetAccessorDeclaration(node: Node): node is GetAccessorDeclaration {
        return node.kind === SyntaxKind.GetAccessor;
    }
```
Both have a decent amount of usage ... `isGetAccessorDeclaration` - 10 usages , `isGetAccessor` - 9 usages, I honestly don't know the difference between them. They are pretty much spread across all compiler layers ..

I will keep `isGetAccessorDeclaration` as other checking methods such as `isMethodDeclaration` and `isPropertyDeclaration` come from `src\compiler\factory\nodeTests.ts` and have no counterpart in `src\compiler\utilitiesPublic.ts`, so it seems more consistent to use all functions from there.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1103,8 +1237,67 @@ namespace ts {
+                    }
+                }
+                else {
+                    return;
```
Does it make sense to `Debug.fail` here?

dragomirtitian replied:
Yes I think so but I think the same applies to the private methods and PR there is a similar `return` for instance members.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1103,8 +1237,67 @@ namespace ts {
             const assignmentExpressions: Expression[] = [];
-
-            if (isPropertyDeclaration(node)) {
+            const env = getPrivateIdentifierEnvironment();
```
There is another `getPrivateIdentifierEnvironment` call a few lines above. We can move this up and access `weakSetName` via `env` below.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -654,10 +759,24 @@ namespace ts {
+                staticProperties = filter(staticProperties, p => !isPrivateIdentifier(p.name));
+            }
+            const staticUninitializedPrivateProperties = !shouldTransformPrivateElements ? []:
+                filter(node.members, (p): p is PropertyDeclaration => isPropertyDeclaration(p) && isPrivateIdentifier(p.name) && hasStaticModifier(p) && !p.initializer);
```
Might be a good idea to pull the predicate into a separate function.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -668,7 +787,7 @@ namespace ts {
             );
 
-            if (some(staticProperties) || some(pendingExpressions)) {
+            if (some(staticUninitializedPrivateProperties) ||some(staticProperties) || some(pendingExpressions)) {
```
```suggestion
            if (some(staticUninitializedPrivateProperties) || some(staticProperties) || some(pendingExpressions)) {

```

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -632,7 +723,21 @@ namespace ts {
+
+            if (languageVersion === ScriptTarget.ESNext) {
+                // We can't initialize static private properties after the class as the are not in the lexical scope
+                // so we rely on standard init even under useDefineForClassFields:false
```
I'm not clear on why `useDefineForClassFields` would be relevant for privates.

dragomirtitian replied:
Actually since we decided emit semantics are broken if we do this, I will just remove this code and emit invalid code under `target:esnext & useDefineForClassFields:false`

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -632,7 +723,21 @@ namespace ts {
+                staticProperties = filter(staticProperties, p => !isPrivateIdentifier(p.name));
+            }
+            else {
+                // We must initialize static private properties even if the don't have a value as we initialize them with a wrapper object which mist always exist.
```
```suggestion
                // We must initialize static private properties even if the don't have a value as we initialize them with a wrapper object which must always exist.

```

mkubilayk replied:
I find this logic a bit counter-intuitive. It looks like initialized and unitialized private static fields should be transformed the same way but because public fields are transformed differently we need do this dance. I left a question below about the wrapper object. If it's not needed, it would simplify some stuff here. If needed, I wonder if we couldn't transform private section in a more uniform way.

dragomirtitian replied:
I think the wrapper object is necessary. Unless we inline the assignment everywhere..
```ts
_C_x = __classStaticPrivateFieldSet(C, C, 10);
```

dragomirtitian committed:
Added error when using initializers with private static fields when useDefineForClassFields is not specified and target is esnext.

dragomirtitian committed:
Fixed code review issues.

dragomirtitian committed:
Removed semantically wrong emit on `useDefineForClassFields:true` with `target:esnext`

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -594,6 +682,9 @@ namespace ts {
         function doesClassElementNeedTransform(node: ClassElement) {
             return isPropertyDeclaration(node) || (shouldTransformPrivateElements && node.name && isPrivateIdentifier(node.name));
         }
+        function getPrivateUnitializedPrivateStaticFields(node: ClassLikeDeclaration) {
```
Says `private` twice in the name.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -603,6 +694,7 @@ namespace ts {
         }
 
         function visitClassDeclaration(node: ClassDeclaration) {
+            getPrivateIdentifierEnvironment().classConstructor = factory.getInternalName(node) ;
```
```suggestion
            getPrivateIdentifierEnvironment().classConstructor = factory.getInternalName(node);

```

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -633,6 +725,15 @@ namespace ts {
             //                                  a lexical declaration such as a LexicalDeclaration or a ClassDeclaration.
             const staticProperties = getProperties(node, /*requireInitializer*/ true, /*isStatic*/ true);
+
+            if (languageVersion < ScriptTarget.ESNext) {
```
This should probably use `shouldTransformPrivateElements` instead.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -633,6 +725,15 @@ namespace ts {
             //      HasLexicalDeclaration (N) : Determines if the argument identifier has a binding in this environment record that was created using
             //                                  a lexical declaration such as a LexicalDeclaration or a ClassDeclaration.
             const staticProperties = getProperties(node, /*requireInitializer*/ true, /*isStatic*/ true);
```
I think I figured what I was trying to express in our meeting today 😅:

- Stop filtering out uninitialized properties here (`requireInitializer = false`).
- Remove `staticUninitializedPrivateProperties` logic below.
- Early return in `transformProperty` for public static uninitialized properties so that they are not transformed. We do that for private uninitialized properties already.
- `createPrivateStaticFieldInitializer` can already handle uninitialized properties.

Does this make sense? This would simply `visitClassExpression`, too.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1341,6 +1534,15 @@ namespace ts {
+        return factory.createAssignment(
+            variableName,
+            factory.createObjectLiteralExpression([
+                factory.createPropertyAssignment("value", initializer || factory.createVoidZero())
```
Should we do the `{ __proto__: null }` trick here? Not really sure if it's worth it.

mkubilayk replied:
I guess it's a "no"? 😄

dragomirtitian replied:
Next on my to do list :)

dragomirtitian replied:
As we discussed we'll skip this at least for now.

dragomirtitian committed:
Changed emit for uninitialized private static fields.

dragomirtitian committed:
Added runtime error in helper if a static private field is accessed before it was declared.

mkubilayk commented on src/compiler/checker.ts:
```diff
@@ -32163,16 +32167,25 @@ namespace ts {
+                            }
+                            // If this is a method/method duplication is might be an overload, so this will be handled when overloads are considered
+                        }
+                        else if (prev & meaning & ~DeclarationMeaning.PrivateStatic) {
```
Just for my own understanding - we ignore the `PrivateStatic` bit here because we don't want private static getters and setters to conflict?

dragomirtitian replied:
Yup that's right.

dragomirtitian replied:
Yup that's right.

mkubilayk commented on src/compiler/checker.ts:
```diff
@@ -32296,6 +32309,9 @@ namespace ts {
                     getNodeLinks(lexicalScope).flags |= NodeCheckFlags.ContainsClassWithPrivateIdentifiers;
                 }
             }
+            if (isPrivateIdentifier(node.name) && hasStaticModifier(node) && node.initializer && languageVersion === ScriptTarget.ESNext && !compilerOptions.useDefineForClassFields) {
```
This line will need to be updated to use `getUseDefineForClassFields()` when https://github.com/microsoft/TypeScript/pull/42663 gets in.

dragomirtitian replied:
Yup. After we get that in we should do a merge and do a sweep for `useDefineForClassFields` again.

dragomirtitian replied:
Yup. After we get that in we should do a merge and do a sweep for `useDefineForClassFields` again. I'll leave this open so we don't forget

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -603,6 +691,7 @@ namespace ts {
         }
 
         function visitClassDeclaration(node: ClassDeclaration) {
+            getPrivateIdentifierEnvironment().classConstructor = factory.getInternalName(node) ;
```
```suggestion
            getPrivateIdentifierEnvironment().classConstructor = factory.getInternalName(node);

```

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -668,7 +766,7 @@ namespace ts {
             );
 
-            if (some(staticProperties) || some(pendingExpressions)) {
+            if (some(staticProperties, p => !!p.initializer || isPrivateIdentifier(p.name)) || some(pendingExpressions)) {
```
Do we need to respect `shouldTransformPrivateElements` here? A simple example like below gets transformed, which doesn't look necessary. I thought it might be because of this line.

```ts
// @target: esnext

const C = class {
    static #x;
}
```

```ts
// @target: esnext
var _a;
const C = (_a = class {
        static #x;
    },
    _a);
```

dragomirtitian replied:
Fixed and I added an extra test for this.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -677,15 +775,19 @@ namespace ts {
 
-                    if (pendingStatements && some(staticProperties)) {
-                        addPropertyStatements(pendingStatements, staticProperties, factory.getInternalName(node));
+                    if (pendingStatements) {
```
nit: I don't think the change here is necessary (going from a single conditional to nested conditionals).

dragomirtitian replied:
Made more sense in a previous version 😅

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1100,11 +1210,69 @@ namespace ts {
             const text = getTextOfPropertyName(node.name) as string;
-            const { weakSetName } = getPrivateIdentifierEnvironment();
+            const env = getPrivateIdentifierEnvironment();
+            const { weakSetName } = env;
```
nit: `const { weakSetName, classConstructor } = env;` would allow us to shorthand property names below.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -1100,11 +1210,69 @@ namespace ts {
+                        info = {
+                            ...previousInfo,
+                            placement: PrivateIdentifierPlacement.StaticGetterAndSetter,
+                            getterName: previousInfo.getterName,
```
We can delete this line. `getterName` is set in spread.

mkubilayk commented on tests/baselines/reference/privateNameAndAny.js:
```diff
@@ -1,8 +1,12 @@
     method(thing: any) {
         thing.#foo; // OK
+        thing.#m();
+        thing.#bar;
```
Is there a reason to duplicate `thing.#bar`?

dragomirtitian replied:
It was supposed to be `#baz`. Thanks.

dragomirtitian replied:
it was supposed to be `#baz`. Fixed

mkubilayk commented on tests/baselines/reference/privateNameDuplicateField.errors.txt:
```diff
@@ -1,114 +1,657 @@
-tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts(47,9): error TS2300: Duplicate identifier '#foo'.
-tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts(48,9): error TS2300: Duplicate identifier '#foo'.
-tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts(49,9): error TS2300: Duplicate identifier '#foo'.
+tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts(6,9): error TS2300: Duplicate identifier '#foo'.
```
Haha what happened here? I this because we replaced the early `return` with `continue` in the checker?

dragomirtitian replied:
Nope. Since duplication detection was so unreliable (combinations of type and order matter)  I deceded to add all possible combinations between instance and static private class elements.

dragomirtitian replied:
Nope. Since duplication detection was so unreliable (combinations of type and order matter) I decided to add all possible combinations between instance and static private class elements.

mkubilayk commented on tests/baselines/reference/privateNameES5Ban(target=es3).errors.txt:
```diff
@@ -18,23 +18,23 @@ tests/cases/conformance/classes/members/privateNames/privateNameES5Ban.ts(10,5):
-        ~~~~~~
-!!! error TS18019: 'static' modifier cannot be used with a private identifier.
+               ~~~~~~~
+!!! error TS18028: Private identifiers are only available when targeting ECMAScript 2015 and higher.
```
I guess technically private static emit can be implemented in earlier versions because it doesn't use `WeakSet` or `WeakMap` but we wan't to have a consistent minimum target?

dragomirtitian replied:
Yes, you are right. I think it would cause more confusion than benefit.

mkubilayk reviewed (commented):
I'm halfway through the fixture changes but reviewed everything else. Leaving a few comments but overall it looks good.

dragomirtitian committed:
Fixed code review comments for private identifier static class elements.

mkubilayk commented on tests/baselines/reference/privateNameStaticFieldUnaryMutation.js:
```diff
@@ -0,0 +1,82 @@
+class C {
+    constructor() {
+        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
+        __classStaticPrivateFieldSet(_a = C, C, _C_test, +__classStaticPrivateFieldGet(_a, C, _C_test) + 1);
+        __classStaticPrivateFieldSet(_b = C, C, _C_test, +__classStaticPrivateFieldGet(_b, C, _C_test) - 1);
+        __classStaticPrivateFieldSet(_c = C, C, _C_test, +__classStaticPrivateFieldGet(_c, C, _C_test) + 1);
```
This is not specific to private statics but it's a bit weird that `C.#test++` and `++C.#test` evaluates to the same thing as standalone expressions. They are correctly handled if the expression is used, though.

mkubilayk commented on tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts:
```diff
@@ -1,53 +1,405 @@
+    // Error
+    class A_StaticGetter_StaticField {
+        static get #foo() { return ""}
+        static #foo() { }
```
```suggestion
        static #foo = "foo";

```

mkubilayk replied:
Sorry, my suggestion is wrong. It should be `static #foo = "foo";`.

mkubilayk reviewed (approved):
Looks good! One small issue in a test.

pushkine committed:
typo (#42899)

* Update editorServices.ts

* Fix baselines

Co-authored-by: Orta <git@orta.io>

Commit:
Fixed typo error. (#42816)

* Fixed three typo errors

* Reverted

* Reverted

* Fixed typo

pushkine committed:
Update project.ts (#42900)

Zzzen committed:
fix(42765): require should not work in ts files (#42792)

* add failing test

* fix(42765): require should not work in ts files

sandersn committed:
Add undefined to Symbol.valueDeclaration (#43033)

* About halfway through the checker

I'm going to merge with master to avoid clashing with the declaration
fix.

* Add undefined to Symbol.valueDeclaration

Also add undefined to a number of utility functions that have always
accepted it, but never added it to their type.

* Fix lint from code review

Co-authored-by: Daniel Rosenwasser <DanielRosenwasser@users.noreply.github.com>

* remove obsoleted fix from inferFromUsage

Co-authored-by: Daniel Rosenwasser <DanielRosenwasser@users.noreply.github.com>

a-tarasyuk committed:
fix(43006): skip trivia in a function name (#43021)

sandersn committed:
Baseline completions and signature help (#43049)

* Initial implementation+first big transitions

* about 10 more

* Change baseline filename + more baselines

1. Use containing file name instead of first @Filename.
2. Switch the rest of the tests I need for @link over to baselines.

* fix lint

* Remove unused/incorrectly named fourslash baselines

* fix incorrectly updated baselines

* dedupe non-unique filenames

* Add names to marker baselines

Also rename another duped test filename.

* Fix semicolon lint

sheetalkamat committed:
Move fixed chunk size polling as a watch option and move it out of server (#42542)

* Move fixed chunk size polling as a watch option and move it out of server
Fixes #41549

* Feedback

sandersn committed:
Delete old, different-case name of quickInfoJSDocTags.baseline (#43052)

armanio123 committed:
Added tests

typescript-bot committed:
Update package-lock.json

dragomirtitian committed:
Merge branch 'es-private-methods-and-accessors' into es-private-static-fields-methods-and-accessors

dragomirtitian committed:
Fixed code review issues for private class elements.

ajafff committed:
fix receiver on calls of imported and exported functions (#35877)

* fix receiver of imported and exported functions

fixes: #35420

* Rebase against master and clean up substitution flow

* Add evaluator tests

* Fix evaluator tests

Co-authored-by: Ron Buckton <ron.buckton@microsoft.com>

armanio123 committed:
Merge pull request #43055 from armanio123/FixDefineForClassFields

Filter transient flags to fix useDefineForClassFields

ajafff committed:
capture thisArg of optionalChaining in parens (#35494)

Fixes: #35476

typescript-bot committed:
Update package-lock.json

DanielRosenwasser committed:
Do a shallow clone for Docker tests. (#42966)

* Do a shallow clone for docker tests.

* Remove the 'pull' step from the Docker files.

sandersn committed:
Update the DOM: March 03, 2021 (#43070)

* Update the DOM: March 03, 2021

Mostly Typescript-DOM-lib-generator#915

* update baselines

rbuckton committed:
Expose 'reservedInNestedScopes' option when creating temp and loop variables (#43083)

a-tarasyuk committed:
fix(39836): allow type declaration/unknown type in catch arguments in JavaScript files (#42392)

DanielRosenwasser committed:
Relax requirements on index signatures to 'any' when a type also contains a string index signature to 'any' (#43065)

* Added test.

* Accepted baselines.

* Allow other index signatures to 'any' if there is a string index signature to 'any'.

* Accepted baselines.

typescript-bot committed:
Update package-lock.json

ahejlsberg committed:
Properly instantiate inferred constraints in conditional types (#42747)

* fix combined type mapper in getConditionalType

* Add regression tests

jonhue committed:
Unnecessary elaboration about not being assignable to type parameters (#42952)

sangxxh committed:
Update parseInt parameter name and jsdoc (#42756)

* feat(lib): improve parseInt type definition and docstring

* Accepted baselines

* update tests

typescript-bot committed:
Update package-lock.json

ahejlsberg committed:
Remove undefined from optional properties when inferring to index signatures (#43086)

* Remove undefined from optional properties when inferring to index signatures

* Add tests

ahejlsberg committed:
Exclude enum/number compatibility rule from comparable relation (#42472)

* Exclude old number/enum literal compatibility rule from comparable relation

* Add tests

* Accept new baselines

typescript-bot committed:
Update package-lock.json

typescript-bot committed:
Update package-lock.json

dragomirtitian committed:
Fixes class shadowing when checking access to a private static class element.

dragomirtitian committed:
Merge remote-tracking branch 'remotes/origin/master' into es-private-static-fields-methods-and-accessors
